rfc/0020-group-policies.md

```diff
@@ -0,0 +1,263 @@
+
+### Expression language
+
+We will use [CEL](https://github.com/google/cel-go) as the expression language for the group policies.
+The main reason for this choice is that CEL is used by the [ValidatingAdmissionPolicy](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/) and [matchConditions](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-matchconditions) in Kubernetes.
+
+Each policy in the group will be represented as a function call in the expression with the same name as the policy defined in the group.
+The expression field should be a valid CEL expression that evaluates to a boolean value and it will be validated by the Kubewarden controller's webhook.
+If the expression evaluates to `true`, the group policy will be considered as `accepted`, otherwise, it will be considered as `rejected`.
+
+### Message
+
+The message field will be used to specify the message that will be returned when the group policy is rejected.
+The specific policy results will be returned in the `warning` field of the response.
+
+This is an example of the response that will be returned when the group policy is rejected:
+
+```json
+{
+  "apiVersion": "admission.k8s.io/v1",
+  "kind": "AdmissionReview",
+  "response": {
+    "uid": "<value from request.uid>",
+    "allowed": false,
+    "warning": [
+      "sigstore_pgp was rejected: the image signature is not valid",
+      "sigstore_gh_action was accepted",
+      "latest_tag was rejected: the image tag latest is not allowed"
+    ]
+  }
+}
+```
+
+### Reconciliation
+
+The new CRDs will be reconciled by the Kubewarden controller.
+The reconciliation flow is similar to the one used for the `AdmissionPolicy` and `ClusterAdmissionPolicy` resources:
+
+- when a new `GroupAdmissionPolicy` or `GroupClusterAdmissionPolicy` is created or updated the reconciler changes the configuration of the policy server to include the created/updated policy group
+- the reconciler rolls out the policy server deployment
+- the reconciler creates or updates the ValidatingWebhookConfiguration pointing to the policy server `validate` endpoint of the policy group
+
+### Match Conditions
+
+An interesting use case is to use the group policies in combination with the [matchConditions](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-matchconditions) feature.
+
+By combining the group policies with `matchConditions`, it could be possible to create complex policies that are evaluated only for specific requests.
+For instance, it could be possible to build a policy that is evaluated only for requests that match a specific label selector.
+
+This RFC does not include any modifications to the policy CRDs to accommodate `matchConditions`, as it falls outside the scope of this proposal.
+
+## Policy server
+
+The policy settings will be extended to include group policies alongside single policies.
```

Comment:
I've used `single` policy all over the RFC, maybe there is a better name for it.

Reply:
Maybe "ordinary" policy or just "policy".

Reply:
I guess you're referring to the `policies.yaml` file used by the Policy Server? If that's correct, I would just put the name of the file in there.

Reply:
I understand it as referring to usual (Cluster)AdmissionPolicies.
I would start naming things as follows: (Cluster)AdmissionPolicyGroups and (Cluster)AdmissionPolicies:

```suggestion
The policy settings will be extended to include PolicyGroups alongside single Policies.
```

Reply:
@viccuad  I've changed to "policy group(s)" . I am avoiding the camel-case as we don't have a resource named `Policy` or `PolicyGroup` (without (Cluster)Admission) and this could be confusing.